std/range/package.d

```diff
@@ -5297,285 +5297,279 @@ auto sequence(alias fun, State...)(State args)
     }
     ---
 */
+auto iota(E)(E end)
+{
+    static if (is(typeof(end += 1)))
+        return IotaResult!(Unqual!E, Unqual!E, int)(E(0), end, 1);
+    else
+        return IotaResult!(Unqual!E, Unqual!E, void)(E(0), end);
+}
+
+
+/// ditto
+auto iota(B, E)(B begin, E end)
+{
+    static if (is(typeof(begin += 1)))
+        return IotaResult!(Unqual!B, Unqual!E, int)(begin, end, 1);
+    else
+        return IotaResult!(Unqual!B, Unqual!E, void)(begin, end);
+}
+
+/// ditto
 auto iota(B, E, S)(B begin, E end, S step)
-if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
-        && isIntegral!S)
 {
-    import std.conv : unsigned;
+    return IotaResult!(Unqual!B, Unqual!E, Unqual!S)(begin, end, step);
+}
+
+struct IotaResult(B, E, S)
+{
+    alias ValueType = CommonType!(B,E);
 
-    alias Value = CommonType!(Unqual!B, Unqual!E);
-    alias StepType = Unqual!S;
+    ValueType current;
+    ValueType end;
 
-    assert(step != 0 || begin == end);
+    static if (isFloatingPoint!ValueType)
+        size_t count, index;
+    static if (hasStep)
+        S step;
 
-    static struct Result
+    enum hasStep = !is(S == void);
+    enum hasLength = is(typeof((end - current) / step) : size_t);
```

Comment:
Such properties should migrate to the top of the type so as to make code easier to follow.